File: src/services/audit/detectors/ad/attack-paths/critical/kerberoasting-to-da.ts
/**
 * Kerberoasting to Domain Admin Path Detector
 *
 * Detects kerberoastable users with path to Domain Admin.
 * Users with SPNs that are members (direct or nested) of privileged groups.
 */

import { ADUser, ADGroup, ADComputer } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';
import { buildGroupMembershipGraph, isPrivilegedMember } from '../../../../../../utils/graph.util';

/**
 * Detect kerberoastable users with path to Domain Admin
 * Users with SPNs that are members (direct or nested) of privileged groups
 */
export function detectPathKerberoastingToDA(
  users: ADUser[],
  groups: ADGroup[],
  computers: ADComputer[],
  includeDetails: boolean
): Finding {
  const graph = buildGroupMembershipGraph(users, groups, computers);

  // Find kerberoastable users (have SPN and are enabled)
  const kerberoastableUsers = users.filter((u) => {
    const spn = u['servicePrincipalName'];
    const hasSPN = spn && Array.isArray(spn) && spn.length > 0;
    return hasSPN && u.enabled;
  });

  // Check which ones have path to privileged groups
  const affected = kerberoastableUsers.filter((u) => isPrivilegedMember(u.dn, graph));

  return {
    type: 'PATH_KERBEROASTING_TO_DA',
    severity: 'critical',
    category: 'attack-paths',
    title: 'Kerberoasting Path to Domain Admin',
    description:
      'User with SPN is member of privileged group. Kerberoasting this account and cracking the password leads to Domain Admin compromise.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
    details:
      affected.length > 0
        ? {
            attackVector: 'Request TGS ticket → Offline crack → Domain Admin access',
            mitigation: 'Use gMSA for service accounts, remove from privileged groups, use long complex passwords',
          }
        : undefined,
  };
}

File: internal/audit/detectors/ad/attack-paths/critical/kerberoasting-to-da.go
package critical

import (
	"context"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// KerberoastingToDADetector detects kerberoastable users with path to Domain Admin
type KerberoastingToDADetector struct {
	audit.BaseDetector
}

// NewKerberoastingToDADetector creates a new detector
func NewKerberoastingToDADetector() *KerberoastingToDADetector {
	return &KerberoastingToDADetector{
		BaseDetector: audit.NewBaseDetector("PATH_KERBEROASTING_TO_DA", audit.CategoryAttackPaths),
	}
}

// Detect executes the detection
func (d *KerberoastingToDADetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.User

	for _, u := range data.Users {
		// Find kerberoastable users (have SPN and are enabled)
		if !u.Enabled() || len(u.ServicePrincipalNames) == 0 {
			continue
		}

		// Check if user is member of privileged groups
		isPrivileged := false
		for _, memberOf := range u.MemberOf {
			memberOfLower := strings.ToLower(memberOf)
			if strings.Contains(memberOfLower, "domain admins") ||
				strings.Contains(memberOfLower, "enterprise admins") ||
				strings.Contains(memberOfLower, "administrators") {
				isPrivileged = true
				break
			}
		}

		if isPrivileged {
			affected = append(affected, u)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityCritical,
		Category:    string(d.Category()),
		Title:       "Kerberoasting Path to Domain Admin",
		Description: "User with SPN is member of privileged group. Kerberoasting this account and cracking the password leads to Domain Admin compromise.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
		finding.Details = map[string]interface{}{
			"attackVector": "Request TGS ticket → Offline crack → Domain Admin access",
			"mitigation":   "Use gMSA for service accounts, remove from privileged groups, use long complex passwords",
		}
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewKerberoastingToDADetector())
}
